### PR TITLE
initial dev to wire in declarative templates into the SDK

### PR DESCRIPTION
**Pull request by bmc-msft**
# Declarative Job Templates

Provide the ability to maintain job templates, akin to `onefuzz template libfuzzer basic` at the service level.

This feature will enable maintaining job templates with user-customizable data alteration.

A declarative job template includes:
* a Job (JobConfig, as used by `onefuzz jobs create`)
* a list of Tasks (TaskConfig, which is used by `onefuzz tasks create`)
* a list of Notifications (NotificationConfig + container type, akin to what is used by `onefuzz notifications create`)
* a list of required and optional form fields used to update the aforementioned JobConfig, TaskConfig, and NotificationConfig entries at runtime.

The form fields allow for 'add' or 'replace' of basic field data using [jsonpatch](http://jsonpatch.com) semantics.

## Example form fields

This following field named `target_workers`, which is required to be an 'int', will optionally (if the request includes it) replace the `target_workers` value of in the first task in the template.

```python
UserField(
    name="target_workers",
    required=False,
    type=UserFieldType.Int,
    locations=[
        UserFieldLocation(
            op=UserFieldOperation.replace,
            path="/tasks/0/task/target_workers",
        ),
    ],
)
```

## Allowed Data Types
As of right now, the data types allowed in configuring arbitrary components in the JobTemplate are:

* bool
* int
* str
* Dict\[str, str\]
* List\[str\]

## Referring to Tasks

The mechanism to refer to pre-existing tasks, such as how `libfuzzer_crash_report` requires `libfuzzer_fuzz` as a prerequisite, is done via specifying the prerequisite task_id.

To support such a reference in `OnefuzzTemplate`, specify the prerequisite TASK by the `u128` representation index in to the list of tasks.  Example, to refer to the first task, use:

```python
TaskConfig(
    prereq_tasks=[UUID(int=0)],
    ...
)
```

## Hardcoded verses Runtime-specified container names

To support differentiating `always use "afl-linux" for tools` verses `ask what container to use for setup`, if the container name is blank in the template, it will be provided as part of the `OnefuzzTemplateConfig` and in the resulting `OnefuzzTemplateRequest`.

## Specifying Notifications in the Template

The existing templates support adding a notification config on the command line, via `--notification_config`, but the existing templates themselves do not include default notifications.

Declarative job templates include optional support to configure notifications as part of the template, rather than requiring the user provide the configuration.

Example declarative job template that specifies using the aforementioned NotificationConfig for the `unique_reports` containers used in the Job.

```python
OnefuzzTemplateNotification(
    container_type=ContainerType.unique_reports,
    notification=NotificationConfig(config=TeamsTemplate(url="https://contoso.com/webhook-url-here")),
)
```

## Current Issues
* Declaratively specifying the allowed values for enums, such as StatsFormat, is not supported.  Fields must currently use Str, which evaluates to Enum value during template rendering, is functional.
* Existing templates automatically differentiate between windows and linux tasks.  This does not support differentiating between platforms automatically.
* Default values are not provided to the user.  

### Implementation details
* Example workflow to create a libfuzzer basic job:
    1. Build a `OnefuzzTemplateConfig` from the specified `OnefuzzTemplate`
    1. SDK uses said `OnefuzzTemplateConfig` plus `argparse` to create a `OnefuzzTemplateRequest`
    1. Render the `OnefuzzTemplate` using the specified `OnefuzzTemplateRequest` 
    1. Use the rendered `OnefuzzTemplate` to create the job, tasks, and notifications
* job\_id in the TaskConfig is can be an arbitrary UUID and is overwritten at
  template evaluation

### Items of note in the implementation
* [_usertemplates.py](https://github.com/bmc-msft/onefuzz/blob/declarative-templates/src/cli/onefuzz/templates/_usertemplates.py): This defines each of the basic templates (such as 'onefuzz template libfuzzer basic')
* [_render_template.py](https://github.com/bmc-msft/onefuzz/blob/declarative-templates/src/cli/onefuzz/templates/_render_template.py) Converts OnefuzzTemplate into OnefuzzTemplateConfig, and renders a OnefuzzTemplate  based on a OnefuzzTemplateRequest
* [_prototype.py](https://github.com/bmc-msft/onefuzz/blob/declarative-templates/src/cli/onefuzz/templates/_prototype.py):  This implements converts OnefuzzTemplateRequest to argparse, and argparse into OnefuzzTemplateRequest, and creates the jobs, tasks, and notifications.

## Output

This is the `OnefuzzTemplateConfig` for `libfuzzer_basic` template, which the CLI will need to turn into an argparse-based wrapper to generate `OnefuzzTemplateRequest` seen below.
```json
{
    "user_fields": [
        {
            "name": "project",
            "type": "Str",
            "required": true
        },
        {
            "name": "name",
            "type": "Str",
            "required": true
        },
        {
            "name": "build",
            "type": "Str",
            "required": true
        },
        {
            "name": "pool_name",
            "type": "Str",
            "required": true
        },
        {
            "name": "target_exe",
            "type": "Str",
            "required": true
        },
        {
            "name": "duration",
            "type": "Int",
            "required": false
        },
        {
            "name": "target_workers",
            "type": "Int",
            "required": false
        },
        {
            "name": "vm_count",
            "type": "Int",
            "required": false
        },
        {
            "name": "target_options",
            "type": "ListStr",
            "required": false
        },
        {
            "name": "target_env",
            "type": "DictStr",
            "required": false
        },
        {
            "name": "reboot_after_setup",
            "type": "Bool",
            "required": false
        },
        {
            "name": "check_retry_count",
            "type": "Int",
            "required": false
        },
        {
            "name": "target_timeout",
            "type": "Int",
            "required": false
        },
        {
            "name": "tags",
            "type": "DictStr",
            "required": false
        }
    ],
    "containers": [
        "inputs",
        "no_repro",
        "readonly_inputs",
        "reports",
        "coverage",
        "setup",
        "unique_reports",
        "crashes"
    ]
}
```

This is the a sample filled in `OnefuzzTemplateRequest` for said form
```json
{
    "template_name": "libfuzzer",
    "user_fields": {
        "project": "my project name",
        "name": "my target name",
        "build": "build # here",
        "pool_name": "windows",
        "target_exe": "fuzz.exe"
    },
    "containers": [
        {
            "type": "no_repro",
            "name": "mynorepro"
        },
        {
            "type": "setup",
            "name": "mysetup"
        },
        {
            "type": "reports",
            "name": "myreports"
        },
        {
            "type": "unique_reports",
            "name": "myuniq"
        },
        {
            "type": "crashes",
            "name": "mycrashes"
        },
        {
            "type": "coverage",
            "name": "mycoverage"
        },
        {
            "type": "inputs",
            "name": "myinputs"
        },
        {
            "type": "readonly_inputs",
            "name": "myinputs"
        }
    ]
}
```

### CODE REVIEW

**Comment by bmc-msft** [comment]
```
$ onefuzz template do_not_use libfuzzer_basic $PLATFORM $PROJECT $TARGET $NAME $POOL --inputs_dir ./seeds/
INFO:onefuzz:creating container: oft-setup-dae851f3b8a75d87a0ddc5006dcf682d
INFO:onefuzz:uploading fuzz.exe to oft-setup-dae851f3b8a75d87a0ddc5006dcf682d
INFO:onefuzz:creating container: oft-no-repro-7d9eddb472f757739d9915650494d665
INFO:onefuzz:creating container: oft-unique-reports-7d9eddb472f757739d9915650494d665
INFO:onefuzz:creating container: oft-reports-7d9eddb472f757739d9915650494d665
INFO:onefuzz:creating container: oft-readonly-inputs-7d9eddb472f757739d9915650494d665
INFO:onefuzz:creating container: oft-crashes-7d9eddb472f757739d9915650494d665
INFO:onefuzz:creating container: oft-coverage-7d9eddb472f757739d9915650494d665
INFO:onefuzz:creating container: oft-inputs-7d9eddb472f757739d9915650494d665
INFO:onefuzz:uploading ./seeds/ to oft-inputs-7d9eddb472f757739d9915650494d665
INFO:onefuzz:created job: f7dca5ea-f0ba-46c7-b700-2a00a17f0928
INFO:onefuzz:creating task. pool:linux type:libfuzzer_fuzz
INFO:onefuzz:creating task. pool:linux type:libfuzzer_crash_report
INFO:onefuzz:creating task. pool:linux type:libfuzzer_coverage
{
    "config": {
        "build": "c",
        "duration": 24,
        "name": "b",
        "project": "a"
    },
    "job_id": "f7dca5ea-f0ba-46c7-b700-2a00a17f0928",
    "state": "init"
}
```

**Comment by bmc-msft** [comment]
All of the user fields should have a "help" field, which is thunked into the argparse help for that item.

File: src/pytypes/onefuzztypes/enums.py [hunk 1]
```diff
@@ -357,3 +361,16 @@ class GithubIssueSearchMatch(Enum):
...
     keep_node_on_completion = "keep_node_on_completion"
+
+
+class UserFieldOperation(Enum):
+    add = "add"
+    replace = "replace"
+
+
+class UserFieldType(Enum):
+    Bool = bool
```
Comment (chkeita) [inline src/pytypes/onefuzztypes/enums.py:372]: might be worth having a guid type

**Comment by bmc-msft** [comment]
For some reason, Github lost association with the repo I originally made this PR from.

I've opened #266 such that I can continue to push updates for this work as I go.  I will port any follow-on comments from this PR into the new one.

**Pull request closed without merging**